connection.go

```diff
@@ -917,11 +940,20 @@ func (conn *Connection) timeouts() {
 			for pos := range shard.requests {
 				shard.rmut.Lock()
 				pair := &shard.requests[pos]
-				for pair.first != nil && pair.first.timeout < nowepoch {
+				currFut := pair.first
+				prevFut := currFut
+				for currFut != nil && currFut.timeout < nowepoch {
 					shard.bufmut.Lock()
-					fut := pair.first
-					pair.first = fut.next
-					if fut.next == nil {
+					if currFut.ctx != nil {
+						prevFut = currFut
+						currFut = currFut.next
+						shard.bufmut.Unlock()
+						continue
+					}
```

Comment:
We can handle two separate lists here:

For requests with timeouts and for requests with contexts. This will be more clear.

Reply:
> This way we need to keep both lists not only here, but also for all functions that handle Future objects lists. If I've understood you right.

Yes, It will help do not waste time in situation: all requests have a context.

Reply:
Done. There are two separated queues now.